File: modules/expo-ssh/android/src/main/java/expo/modules/ssh/SSHSession.kt
package expo.modules.ssh

import net.schmizz.sshj.DefaultConfig
import net.schmizz.sshj.SSHClient
import net.schmizz.sshj.common.Buffer
import net.schmizz.sshj.common.Ed25519KeyFactory
import net.schmizz.sshj.connection.channel.direct.Session
import net.schmizz.sshj.sftp.FileAttributes
import net.schmizz.sshj.sftp.OpenMode
import net.schmizz.sshj.transport.verification.HostKeyVerifier
import net.schmizz.sshj.userauth.keyprovider.KeyPairWrapper
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.jce.provider.BouncyCastleProvider
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.security.KeyPair
import java.security.PublicKey
import java.security.Security
import java.util.EnumSet
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread

/**
 * The sshj wrapper: connect, ed25519 auth, one PTY, short-lived exec channels, SFTP. The same
 * behavioural contract as `ios/SSHSession.swift`; nothing above this sees an sshj type.
 *
 * Library choice (AGENTS.md: name what you looked at) — **sshj 0.40.0** over Apache MINA
 * sshd-client, both read at source before deciding:
 * - sshj covers every requirement in-tree, verified at tag v0.40.0: `HostKeyVerifier.verify` is a
 *   synchronous boolean callback we can block until JS answers the TOFU prompt;
 *   `Ed25519KeyFactory.getPrivateKey` takes exactly our raw 32-byte seed;
 *   `SessionChannel` has `allocatePTY(term, cols, rows, …)`, `startShell`,
 *   `changeWindowDimensions` and `exec` with an exit status; `SFTPEngine.makeDir(path, attrs)`
 *   takes the 0700 mode that `SFTPClient.mkdir` drops; `RemoteFile.write` writes bytes at an
 *   offset; `RemoteResourceInfo` carries name/type/size.
 * - MINA sshd covers the list too, but as an async server+client framework — its own IO reactor,
 *   futures graph, a multi-artifact split, and ed25519 only through *optional* dependencies
 *   (docs/dependencies.md at sshd-2.15.0). A lot of machinery for one blocking connection per
 *   app. sshj is client-only, blocking and small; sshj's own test suite runs MINA as the
 *   throwaway server.
 */
class SSHSession {
  companion object {
    init {
      // Android registers a stripped BouncyCastle under the same "BC" name (no Ed25519
      // KeyFactory), and sshj's SecurityUtils resolves algorithms through that name. Swap in the
      // full bcprov this module bundles; appended last, so nothing else's provider order moves.
      if (Security.getProvider(BouncyCastleProvider.PROVIDER_NAME)?.javaClass != BouncyCastleProvider::class.java) {
        Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME)
        Security.addProvider(BouncyCastleProvider())
      }
    }
  }

  data class RemoteEntry(val name: String, val isDirectory: Boolean, val size: Double)

  @Volatile private var client: SSHClient? = null
  @Volatile private var shellSession: Session? = null
  @Volatile private var shell: Session.Shell? = null

  private val hostKeyLock = Object()
  private var hostKeyAnswer: Boolean? = null

  /** `onHostKey` receives the host key in SSH wire format — the bytes the JS side fingerprints
   *  and pins. It fires on sshj's transport thread mid-handshake, and the handshake then blocks
   *  in [awaitHostKeyDecision] until [resolveHostKey] answers — that block is what holds the
   *  connection open while a human reads the fingerprint. */
  fun connect(host: String, port: Int, username: String, seed: ByteArray, onHostKey: (ByteArray) -> Unit) {
    synchronized(hostKeyLock) { hostKeyAnswer = null }
    val keyPair = keyPairFromSeed(seed)
    val client = SSHClient(DefaultConfig())
    client.addHostKeyVerifier(object : HostKeyVerifier {
      override fun verify(hostname: String, port: Int, key: PublicKey): Boolean {
        onHostKey(Buffer.PlainBuffer().putPublicKey(key).compactData)
        return awaitHostKeyDecision()
      }

      override fun findExistingAlgorithms(hostname: String, port: Int): List<String> = emptyList()
    })
    try {
      client.connect(host, port)
      client.authPublickey(username, KeyPairWrapper(keyPair))
    } catch (failure: Exception) {
      runCatching { client.disconnect() }
      throw failure
    }
    this.client = client
  }

  /** sshj's own factory wraps a 32-byte seed into the PKCS#8 form the JCA wants; the public half
   *  comes from BouncyCastle's lightweight curve math, which bcprov already ships. */
  private fun keyPairFromSeed(seed: ByteArray): KeyPair {
    val publicKey = Ed25519PrivateKeyParameters(seed, 0).generatePublicKey().encoded
    return KeyPair(Ed25519KeyFactory.getPublicKey(publicKey), Ed25519KeyFactory.getPrivateKey(seed))
  }

  // MARK: - Host key answer

  /** The answer can beat the question: a pinned key is matched in JS the moment `onHostKey`
   *  lands, which can be before the verifier blocks here. So an early answer is kept, not
   *  dropped — same contract as the iOS actor. */
  private fun awaitHostKeyDecision(): Boolean {
    synchronized(hostKeyLock) {
      try {
        while (hostKeyAnswer == null) hostKeyLock.wait()
      } catch (_: InterruptedException) {
        return false
      }
      val answer = hostKeyAnswer == true
      hostKeyAnswer = null
      return answer
    }
  }

  fun resolveHostKey(accept: Boolean) {
    synchronized(hostKeyLock) {
      hostKeyAnswer = accept
      hostKeyLock.notifyAll()
    }
  }

  // MARK: - PTY

  /** Opens the one PTY this app ever needs. Output lands in [onData] until the channel dies,
   *  then [onClose] fires once — remote hangup and our own [disconnect] look the same. */
  fun startShell(cols: Int, rows: Int, term: String, onData: (ByteArray) -> Unit, onClose: () -> Unit) {
    val client = client ?: throw IllegalStateException("Not connected")
    val session = client.startSession()
    session.allocatePTY(term, cols, rows, 0, 0, emptyMap())
    val shell = session.startShell()
    shellSession = session
    this.shell = shell
    // With a PTY there is no separate stderr; everything the remote prints is on this one stream.
    thread(name = "expo-ssh-shell", isDaemon = true) {
      val input = shell.inputStream
      val chunk = ByteArray(32 * 1024)
      try {
        while (true) {
          val read = input.read(chunk)
          if (read < 0) break
          if (read > 0) onData(chunk.copyOf(read))
        }
      } catch (_: IOException) {
        // A dead channel is a finished stream; the JS reconnect state machine decides what next.
      }
      onClose()
    }
  }

  fun send(text: String) {
    val shell = shell ?: throw IllegalStateException("Not connected")
    val output = shell.outputStream
    output.write(text.toByteArray(Charsets.UTF_8))
    output.flush() // ChannelOutputStream buffers until flushed
  }

  fun resize(cols: Int, rows: Int) {
    val shell = shell ?: throw IllegalStateException("Not connected")
    shell.changeWindowDimensions(cols, rows, 0, 0)
  }

  // MARK: - Exec

  /** Runs one command on its own short-lived exec channel and returns what it printed.
   *
   *  Nothing here touches the PTY: the tmux side-channel commands must not reach the attached
   *  shell, which would echo them into the grid the user is looking at. Capped, because
   *  `capture-pane` on a pane with a long line is unbounded otherwise. A non-zero exit throws,
   *  like Citadel's `executeCommand` on iOS — the callers treat that the same as empty output. */
  fun exec(command: String, limit: Int): String {
    val client = client ?: throw IllegalStateException("Not connected")
    client.startSession().use { session ->
      val channel = session.exec(command)
      val output = readCapped(channel.inputStream, limit)
      channel.join()
      val status = channel.exitStatus
      if (status != null && status != 0) throw IOException("Command exited $status")
      return String(output, Charsets.UTF_8)
    }
  }

  private fun readCapped(input: InputStream, limit: Int): ByteArray {
    val collected = ByteArrayOutputStream()
    val chunk = ByteArray(8 * 1024)
    while (true) {
      val read = input.read(chunk)
      if (read < 0) break
      // Past the cap the stream still drains, so the channel closes and hands over the exit code.
      val keep = minOf(read, limit - collected.size())
      if (keep > 0) collected.write(chunk, 0, keep)
    }
    return collected.toByteArray()
  }

  /** Whether the far end is still there — asked, not assumed. `isConnected` is local channel
   *  state, so a half-open TCP (phone slept, Wi-Fi roamed, no FIN ever arrives) reads healthy
   *  forever. A round trip is the only test a half-open socket cannot fake: it swallows the
   *  request and says nothing, and saying nothing is the answer. */
  fun isAlive(timeoutMs: Long): Boolean {
    if (client?.isConnected != true) return false
    // ponytail: a timed-out probe leaves its pool thread blocked until the connection dies —
    // same as the iOS raced task; a cancellable channel would be the upgrade if it ever matters.
    val roundTrip = CompletableFuture.supplyAsync { runCatching { exec("true", 64) }.isSuccess }
    return try {
      roundTrip.get(timeoutMs, TimeUnit.MILLISECONDS)
    } catch (_: Exception) {
      false
    }
  }

  // MARK: - SFTP

  /** One file onto the host over the connection's SFTP subsystem. Off the PTY, like [exec]:
   *  nothing here is echoed into the grid.
   *
   *  `directories` is the mkdir chain for the path, shallowest first — SFTP mkdir has no `-p`.
   *  Each level is tried and ignored: it fails whenever it is already there, and SFTP has no test
   *  that is cheaper than trying. `SFTPClient.mkdir` drops the mode, hence the engine call. */
  fun upload(data: ByteArray, path: String, directories: List<String>) {
    val client = client ?: throw IllegalStateException("Not connected")
    client.newSFTPClient().use { sftp ->
      val mode = FileAttributes.Builder().withPermissions(448).build() // 0700
      for (directory in directories) {
        runCatching { sftp.sftpEngine.makeDir(directory, mode) }
      }
      sftp.open(path, EnumSet.of(OpenMode.WRITE, OpenMode.CREAT, OpenMode.TRUNC)).use { file ->
        // One WRITE packet per call — sshj's RemoteFile.write does not chunk, so we do, at the
        // 32 KB every server accepts.
        var offset = 0
        while (offset < data.size) {
          val length = minOf(32 * 1024, data.size - offset)
          file.write(offset.toLong(), data, offset, length)
          offset += length
        }
      }
    }
  }

  /** The listing behind the upload destination browser. Read-only: this module never removes or
   *  downloads anything. sshj keys `isDirectory` off the readdir attributes and never exposes
   *  `longname`, so the iOS fallback for servers that omit attributes has no equivalent here. */
  fun listDirectory(path: String): List<RemoteEntry> {
    val client = client ?: throw IllegalStateException("Not connected")
    client.newSFTPClient().use { sftp ->
      return sftp.ls(path).map {
        RemoteEntry(it.name, it.isDirectory, it.attributes.size.toDouble())
      }
    }
  }

  // MARK: - Teardown

  fun disconnect() {
    resolveHostKey(false) // unblocks a handshake still waiting on the prompt
    runCatching { shellSession?.close() }
    runCatching { client?.disconnect() }
    shell = null
    shellSession = null
    client = null
  }
}

File: modules/expo-ssh/ios/SSHSession.swift
// Curve25519.Signing.PrivateKey is not Sendable, and Citadel's auth method is a @Sendable closure.
@preconcurrency import Crypto
import Foundation
// Citadel's own dependency — `withSFTP` takes a Logger, and the default one prints the paths it opens.
import Logging
import NIOCore
import NIOSSH

// Citadel's TTYStdinWriter and TTYOutput predate strict concurrency and are not marked Sendable.
@preconcurrency import Citadel

/// The Citadel wrapper: connect, ed25519 auth, one PTY, short-lived exec channels, SFTP.
/// Ported from the reference app's `Port22Core/SSHSession.swift`; nothing above this sees a NIO type.
actor SSHSession {
  enum Failure: Error {
    case notConnected
    case hostKeyRejected
    case badPayload
  }

  struct RemoteEntry: Sendable {
    let name: String
    let isDirectory: Bool
    let size: Int
  }

  private var client: SSHClient?
  private var writer: TTYStdinWriter?
  private var ptyTask: Task<Void, Never>?
  private var ptyWaiters: [CheckedContinuation<Void, Never>] = []
  private var hostKeyWaiter: CheckedContinuation<Bool, Never>?
  private var hostKeyAnswer: Bool?

  /// `verify` receives the host key in SSH wire format — the bytes the JS side fingerprints and pins.
  /// Returning false aborts the handshake. It is async because the answer may be a human tapping
  /// Trust, and the handshake has to wait for that without parking an event loop thread.
  func connect(
    host: String,
    port: Int,
    username: String,
    seed: Data,
    verify: @escaping @Sendable (Data) async -> Bool
  ) async throws {
    let privateKey = try Curve25519.Signing.PrivateKey(rawRepresentation: seed)
    hostKeyAnswer = nil
    client = try await SSHClient.connect(
      to: SSHClientSettings(
        host: host,
        port: port,
        authenticationMethod: { .ed25519(username: username, privateKey: privateKey) },
        hostKeyValidator: .custom(HostKeyCallback(verify: verify))
      )
    )
  }

  // MARK: - Host key answer

  /// The answer can beat the question: a pinned key is matched in JS the moment `onHostKey` lands,
  /// which can be before the handshake gets here. So an early answer is kept rather than dropped.
  func awaitHostKeyDecision() async -> Bool {
    if let answer = hostKeyAnswer {
      hostKeyAnswer = nil
      return answer
    }
    return await withCheckedContinuation { hostKeyWaiter = $0 }
  }

  func resolveHostKey(_ accept: Bool) {
    guard let waiter = hostKeyWaiter else {
      hostKeyAnswer = accept
      return
    }
    hostKeyWaiter = nil
    waiter.resume(returning: accept)
  }

  // MARK: - PTY

  /// Opens the one PTY this app ever needs and streams its output. The stream finishes when the
  /// remote side closes the channel or `disconnect()` runs.
  func requestPTY(cols: Int, rows: Int, term: String) async throws -> AsyncStream<[UInt8]> {
    guard let client else { throw Failure.notConnected }

    let request = SSHChannelRequestEvent.PseudoTerminalRequest(
      wantReply: true,
      term: term,
      terminalCharacterWidth: cols,
      terminalRowHeight: rows,
      terminalPixelWidth: 0,
      terminalPixelHeight: 0,
      terminalModes: SSHTerminalModes([:])
    )

    let (stream, continuation) = AsyncStream<[UInt8]>.makeStream()

    // Citadel scopes a PTY to a closure, so the closure stays parked for the life of the session.
    // The writer it hands out is stashed on the actor; the task ends when the channel does.
    ptyTask = Task { [weak self] in
      do {
        try await client.withPTY(request) { inbound, outbound in
          await self?.ptyOpened(outbound)
          for try await output in inbound {
            switch output {
            case .stdout(let buffer), .stderr(let buffer):
              continuation.yield(Array(buffer.readableBytesView))
            }
          }
        }
      } catch {
        // A dead channel is a finished stream; the JS reconnect state machine decides what next.
      }
      continuation.finish()
      await self?.ptyClosed()
    }

    await waitForPTY()
    guard writer != nil else { throw Failure.notConnected }
    return stream
  }

  func send(_ text: String) async throws {
    guard let writer else { throw Failure.notConnected }
    try await writer.write(ByteBuffer(string: text))
  }

  func resize(cols: Int, rows: Int) async throws {
    guard let writer else { throw Failure.notConnected }
    try await writer.changeSize(cols: cols, rows: rows, pixelWidth: 0, pixelHeight: 0)
  }

  // MARK: - Exec

  /// Runs one command on its own short-lived exec channel and returns what it printed.
  ///
  /// Nothing here touches the PTY: the tmux side-channel commands must not reach the attached
  /// shell, which would echo them into the grid the user is looking at. Capped, because
  /// `capture-pane` on a pane with a long line is unbounded otherwise.
  func run(_ command: String, limit: Int) async throws -> String {
    guard let client else { throw Failure.notConnected }
    let output = try await client.executeCommand(command, maxResponseSize: limit)
    return String(decoding: output.readableBytesView, as: UTF8.self)
  }

  /// Whether the far end is still there — asked, not assumed. Citadel's `isConnected` is local
  /// channel state, so a half-open TCP (phone slept, Wi-Fi roamed, no FIN ever arrives) reads
  /// healthy forever. A round trip is the only test a half-open socket cannot fake: it swallows
  /// the request and says nothing, and saying nothing is the answer.
  func isAlive(timeout: Duration) async -> Bool {
    guard client != nil else { return false }
    return await withTaskGroup(of: Bool.self) { group in
      group.addTask { (try? await self.run("true", limit: 64)) != nil }
      group.addTask {
        try? await Task.sleep(for: timeout)
        return false
      }
      let answered = await group.next() ?? false
      group.cancelAll()
      return answered
    }
  }

  // MARK: - SFTP

  /// One file onto the host over the connection's SFTP subsystem. Off the PTY, like `run`: nothing
  /// here is echoed into the grid. An exec channel could not do this at all — `executeCommandStream`
  /// exposes stdout and stderr only, so there is no stdin to pipe the bytes into.
  ///
  /// `directories` is the mkdir chain for the path, shallowest first — SFTP mkdir has no `-p`. Each
  /// one is tried and ignored: it fails whenever the level is already there, and SFTP has no test
  /// that is cheaper than trying.
  func upload(_ data: [UInt8], to path: String, creating directories: [String]) async throws {
    guard let client else { throw Failure.notConnected }
    try await client.withSFTP(logger: Self.chatty) { sftp in
      var attributes = SFTPFileAttributes()
      attributes.permissions = 0o700
      for directory in directories {
        try? await sftp.createDirectory(atPath: directory, attributes: attributes)
      }
      try await sftp.withFile(filePath: path, flags: [.write, .create, .truncate]) { file in
        // Citadel chunks this at 32 KB itself; nothing here has to know the message limit.
        try await file.write(ByteBuffer(bytes: data))
      }
    }
  }

  /// The listing behind the upload destination browser. Read-only: this module never removes or
  /// downloads anything.
  func listDirectory(_ path: String) async throws -> [RemoteEntry] {
    guard let client else { throw Failure.notConnected }
    return try await client.withSFTP(logger: Self.chatty) { sftp in
      try await sftp.listDirectory(atPath: path).flatMap(\.components).map { component in
        RemoteEntry(
          name: component.filename,
          isDirectory: Self.isDirectory(component),
          size: Int(component.attributes.size ?? 0)
        )
      }
    }
  }

  /// `permissions` is optional in the protocol; `longname` is `ls -l` output, whose first character
  /// is the file type, and every server that omits attributes still sends it.
  private static func isDirectory(_ component: SFTPPathComponent) -> Bool {
    if let permissions = component.attributes.permissions {
      return permissions & 0xF000 == 0x4000  // S_IFDIR
    }
    return component.longname.hasPrefix("d")
  }

  /// Citadel's SFTP client announces every path it opens at `.info`. That used to be silenced;
  /// PLAN.md §7 now says log freely, so it talks. This lands in the device console rather than
  /// Metro (`idevicesyslog`), and it is the only view into an upload that fails inside the
  /// subsystem rather than at the call. `SFTPFile` inherits it.
  private static var chatty: Logger {
    var logger = Logger(label: "port22.sftp")
    logger.logLevel = .debug
    return logger
  }

  // MARK: - Teardown

  func disconnect() async {
    resolveHostKey(false)
    try? await client?.close()
    ptyTask?.cancel()
    await ptyTask?.value
    ptyTask = nil
    writer = nil
    client = nil
  }

  // MARK: - PTY readiness

  private func ptyOpened(_ writer: TTYStdinWriter) {
    self.writer = writer
    resumeWaiters()
  }

  private func ptyClosed() {
    writer = nil
    resumeWaiters()
  }

  private func resumeWaiters() {
    let waiters = ptyWaiters
    ptyWaiters = []
    for waiter in waiters { waiter.resume() }
  }

  private func waitForPTY() async {
    guard writer == nil else { return }
    await withCheckedContinuation { ptyWaiters.append($0) }
  }
}

/// Bridges Citadel's host key check to the JS callback. It hands over the wire-format key and
/// nothing else — no logging, ever.
private struct HostKeyCallback: NIOSSHClientServerAuthenticationDelegate {
  let verify: @Sendable (Data) async -> Bool

  func validateHostKey(hostKey: NIOSSHPublicKey, validationCompletePromise: EventLoopPromise<Void>) {
    var buffer = ByteBuffer()
    hostKey.write(to: &buffer)
    let key = Data(buffer.readableBytesView)
    // Off the event loop: the answer may be a human tapping Trust.
    Task {
      if await verify(key) {
        validationCompletePromise.succeed(())
      } else {
        validationCompletePromise.fail(SSHSession.Failure.hostKeyRejected)
      }
    }
  }
}
